### PR TITLE
[nativert] Move Pytree

### PR DESCRIPTION
**Pull request by yushangdi**
Summary: fbcode/sigmoid/core/common -> fbcode/caffe2/torch/nativert/common

Torch Native Runtime RFC: https://github.com/pytorch/rfcs/pull/72

Test Plan:
```
buck run fbcode//mode/dev-nosan  //caffe2/test/cpp/nativert:pytree_test
```

OSS CI

Rollback Plan:

Differential Revision: D75965059

### CODE REVIEW

**Comment by Skylion007** [comment]
I'm sure @XuehaiPan has some thoughts here

**Comment by zhxchen17** [comment]
I think we mean to move the code to nativert/ as part of the impl details. To add some context, we don't intend to expose this as part of public API. Also this pytree impl contains many optimizations we accumulated to high qps low latency model serving use cases and imo we should still recommend ppl to use optree in common cases.

**Comment by yushangdi** [comment]
> Also this pytree impl contains many optimizations we accumulated to high qps low latency model serving use cases and imo we should still recommend ppl to use optree in common cases.

@zhxchen17 sure, I'll move it to `nativert/detail` instead then.

**Review by zhxchen17 — APPROVED** [review]

**Review by albanD — DISMISSED** [review]
@zou3519 how does that align with the other c++ pytree discussion we've been having?

**Comment by zou3519** [comment]
> @zou3519 how does that align with the other c++ pytree discussion we've been having?

@albanD when we say C++ pytree we really mean two things:
1) Python bindings for C++ functionality to flatten/unflatten PyObject* Python containers
2) C++ functionality to flatten/unflatten IValues

"C++ Pytree" and optree is (1). We don't have (2) in PyTorch. This PR adds (2).
There are use cases for both, e.g. executorch has both (1) and (2).

My conclusion is that we should allow this PR and that it is unrelated to the optree story. If we're up to it then we should probably name it something else other than "pytree", because there's no Python involved

**Comment by yushangdi** [comment]
>  If we're up to it then we should probably name it something else other than "pytree", because there's no Python involved

I'd argue it's not completely unrelated to python, because it does register and process the string representation of python types in json, e.g. `builtins.tuple`, `torch.fx.immutable_collections.immutable_dict`...

But if we really want a name other than pytree, @zou3519 @zhxchen17  do you have any suggestion for a better name? maybe `TreeSpec.h`?

File: torch/nativert/detail/Pytree.h [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+namespace torch::nativert::detail {
+
+class TreeSpec;
+
+using TreeFlattenFn =
+    void (*)(const c10::IValue&, const TreeSpec&, std::vector<c10::IValue>&);
+using TreeUnflattenFn =
+    c10::IValue (*)(std::vector<c10::IValue>, const nlohmann::json&);
+
+using ContextLoadFn = nlohmann::json (*)(std::string_view);
```
Comment (XuehaiPan) [inline torch/nativert/detail/Pytree.h]: Since the context is limited to JSON objects or strings, I'm sure this "pytree" utility cannot easily support arbitrary Python container types.

If I understand correctly, this is a C++ extension bridged with the Python pytree serialization/unserialization format for `torch.export`.

------

> But if we really want a name other than pytree, @zou3519 @zhxchen17 do you have any suggestion for a better name? maybe TreeSpec.h?

I previously tried to implement a draft for C++ pytree in `libtorch` named `torch::utils::nested`.

- #146893

FYI, TensorFlow has a module that handles nested Python containers, called [`tf.nest`](https://www.tensorflow.org/api_docs/python/tf/nest). It only has limited support with native container types in Python stdlib, and does not support registering custom types.

File: torch/nativert/detail/Pytree.cpp [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+    if (children_.empty()) {
+      return c10::DictType::create(c10::AnyType::get(), c10::AnyType::get());
+    } else {
+      return c10::DictType::create(
+          dynamicToIValue(context_[0]).type(), children_[0].toAtenType());
+    }
+  } else {
+    TORCH_CHECK(false, "Unsupported uniform name: ", uniformName());
+  }
+}
```
Comment (XuehaiPan) [inline torch/nativert/detail/Pytree.cpp]: We should support `PyStructSequence` types in `torch.return_types.*`.

File: torch/nativert/detail/Pytree.h (hunk 1 above)
Comment (zou3519) [reply torch/nativert/detail/Pytree.h]: Hmm you're right, the way I would describe it is that it's a bridge. Given some treespec in Python, and some Python flat_args, this package allows us to take C++ IValue flat_args and use a converted version of the treespec with it.

Yes, I would describe it as similar to #146893

File: torch/nativert/detail/Pytree.cpp (hunk 2 above)
Comment (yushangdi) [reply torch/nativert/detail/Pytree.cpp]: > We should support `PyStructSequence` types in `torch.return_types.*`.

We can add this as a followup. This is mainly used for the serialized input and output specs of exported program. Currently, there's no default serialization implemented for `torch.return_types.*`. The users would need to update their _register_pytree_node call with a `serialized_type_name` kwarg for these types. If we want to hard-code the  `uniformName` here for `torch.return_types.*`, it might be good to do that along with a default serialization implementation for exported program. 

cc @zhxchen17

Comment (XuehaiPan) [reply torch/nativert/detail/Pytree.cpp]: > If we want to hard-code the `uniformName` here for `torch.return_types.*`

`torch.return_types.*` are first-party types that we can hardcode the behaviors.

Comment (yushangdi) [reply torch/nativert/detail/Pytree.cpp]: > > If we want to hard-code the `uniformName` here for `torch.return_types.*`
> 
> `torch.return_types.*` are first-party types that we can hardcode the behaviors.

@XuehaiPan yeah, I'm just saying if we hard code it here, we need to hard code it in the exported program serializer as well. And I think it might be good to add and test this in a separate PR. In this PR we're just moving the code to OSS. We can extend the code as a follow up.

**Comment by yushangdi** [comment]
@XuehaiPan @zou3519 How about we just re-use the `NestedStructure.h` name from  https://github.com/pytorch/pytorch/pull? I can change the file name from `Pytree.h` and `Pytree.cpp` to  `NestedStructure.h` and `NestedStructure.cpp`.

**Comment by XuehaiPan** [comment]
+1 for changing `Tree` -> `Nested`, `Leaf` -> `Atom`/`IValue`.

Another thing I want to point out is that when we call with pure-C containers, we should take care of the recursion limit. We will encounter vulnerabilities when a cycle exists in the nested structure.

```cpp
Flatten(...) {
    return FlattenHelper(..., /*depth=*/0);
}

FlattenHelper(..., size_t depth) {
    // recursion code
}
```

File: torch/nativert/detail/Pytree.cpp (hunk 2 above)
Comment (yushangdi) [reply torch/nativert/detail/Pytree.cpp]: @XuehaiPan I'm adding a name for the torch.return_types in https://github.com/pytorch/pytorch/pull/155245 so we can pytree dump them. Once that's done, I can add them here in a follow up PR.

**Review by XuehaiPan — APPROVED** [review]

**Comment by yushangdi** [comment]
@albanD Can I have a review again? Thanks!

File: torch/nativert/detail/NestedStructure.h [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    const NestedSpec& spec);
+
+std::vector<c10::IValue> nestedFlatten(
+    const c10::IValue& nested,
+    const NestedSpec& spec);
+
+std::vector<c10::IValue> nestedFlattenFromArgs(
+    const std::vector<c10::IValue>& args,
+    const std::unordered_map<std::string, c10::IValue>& kwargs,
+    const NestedSpec& spec);
```
Comment (zou3519) [inline torch/nativert/detail/NestedStructure.h]: If you don't mind, I think I want to bikeshed the name some more, especially because this is going into PyTorch proper.

1) I think it's OK to call the data structure a TreeSpec. TreeSpec doesn't imply any python-ness. You already use the word "NodeDef" here.
2) Should we remove the nested prefix from all of the APIs? Flatten, FlattenToTensorList, etc seem fine to have. If we do want a prefix, then i would still prefer "Tree" as the prefix.
3) Instead of NestedStructure, this is more like... a C++ Tree data structure. pytree = python tree, cpptree = cpp tree, C++ pytree = C++-backed implementation of python tree. I'd prefer us name this "cpptree" or something.

Thoughts, @XuehaiPan @yushangdi ?

**Review by zou3519 — COMMENTED** [review]
I want to bikeshed the naming more but otherwise this looks fine to me.

File: torch/nativert/detail/NestedStructure.h (hunk 3 above)
Comment (XuehaiPan) [reply torch/nativert/detail/NestedStructure.h]: I suggest using "nested" since the "tree" term has its own meaning in C++ (e.g., `std::set`, `std::map`).

If we have a strong preference to use "tree", I'd like to use `ITree` corresponding to `IValue` rather than cpptree.

Comment (zou3519) [reply torch/nativert/detail/NestedStructure.h]: How is "tree" used in std::set and std::map? Like a red-black tree?

The main reason why I want to use "tree" is for consistency with the python and C++ APIs. We could have named the python API "nestedstructure" too but we didn't.

Comment (zhxchen17) [reply torch/nativert/detail/NestedStructure.h]: ITree or IValueTree both sounds okay to me. I'm fine with NestedStructure as well but I think it will cause more confusion when people read about these and wonder what's the difference between this and pytree while conceptually they look pretty similar.

Comment (XuehaiPan) [reply torch/nativert/detail/NestedStructure.h]: > How is "tree" used in std::set and std::map? Like a red-black tree?

The underlying implementation of `std::set` and `std::map` is the red-black tree. For the term _'tree'_ typically used in C++, there is often a value associated with each node (including internal nodes and terminal nodes).

```cpp
template <typename T>
struct Node {
    T value;
    Node* children;
};
```

For `pytree`, it comes from JAX pytree, where the concept is borrowed from the DM-Tree package (https://github.com/google-deepmind/tree). Since the module name `tree` is already taken, a `py-` prefix is added to avoid name shadowing.

In DM-Tree, they use _'structure'_ to refer to their data structure. See https://tree.readthedocs.io/en/latest/api.html

```cpp
Structure = Union[
    Any,
    Sequence['Structure'],
    Mapping[Any, 'Structure'],
    'AnyNamedTuple',
]

tree.map_structure(func, structure)
tree.is_nested(structure)
tree.flatten(structure)
tree.unflatten_as(structure, flat_sequence)
```

The in-order, pre-order, and post-order traversals of a pytree are equal to each other. Because the values are only associated with the terminal nodes.

------

> The main reason why I want to use "tree" is for consistency with the python and C++ APIs.

I'm fine with this. See my comment above, I'd prefer to avoid using the `cpp-` prefix. I vote for `ITree` or `IValueTree` for:

```txt
 C++             Python

IValue    <---> PyObject
ITree     <---> PyTree
ITreeSpec <---> PyTreeSpec
```

Comment (yushangdi) [reply torch/nativert/detail/NestedStructure.h]: @zou3519 @XuehaiPan @zhxchen17 To summarize the discussion, how about the following? 

- Use IValue/ITree/ITreeSpec name for classes/structs
- Remove nested prefix from all of the function APIs, e.g. Flatten, FlattenToTensorList...

Comment (zou3519) [reply torch/nativert/detail/NestedStructure.h]: Using "I" as the prefix sounds good to me, this is an IValue-based flatten/unflatten API

Comment (yushangdi) [reply torch/nativert/detail/NestedStructure.h]: > Using "I" as the prefix sounds good to me, this is an IValue-based flatten/unflatten API

@zou3519 Just to double check, do you want the I-prefix to also exist on the function names? e.g. `IFlatten` or just `Flatten`, `IFlattenToTensorList` or `FlattenToTensorList`. I have a slight preference on `IFlatten` since this is less likely to have name collision (`Flatten` might be a common name).

Comment (zou3519) [reply torch/nativert/detail/NestedStructure.h]: I don't have a preference for the function names, up to you

**Comment by yushangdi** [comment]
@zou3519 updated to use `ITree` now :)

**Review by zou3519 — APPROVED** [review]
